tests/python/relax/test_transform_extract_dataflow_blocks.py

```diff
@@ -0,0 +1,268 @@
+                with R.dataflow():
+                    q = R.add(v, v)
+                    a = R.multiply(q, q)
+                    b = R.add(a, a)
+                    R.output(b)
+                c = b
+            with R.dataflow():
+                d = R.multiply(c, c)
+                e = R.add(d, d)
+                f = R.multiply(e, e)
+                R.output(f)
+            return f
+
+
+class TestTreatNonCallAsPure(ExtractCompare):
+    @I.ir_module
+    class Before:
+        @R.function
+        def main(t: R.Tuple(R.Tensor, R.Tensor)) -> R.Tensor:
+            x = t[0]
+            y = t[1]
+            z = R.add(x, y)
+            w = R.multiply(z, z)
+            return w
+
+    @I.ir_module
+    class Expected:
+        @R.function
+        def main(t: R.Tuple(R.Tensor, R.Tensor)) -> R.Tensor:
+            with R.dataflow():
+                x = t[0]
+                y = t[1]
+                z = R.add(x, y)
+                w = R.multiply(z, z)
+                R.output(w)
+            return w
+
+
+class TestInnerFunction(ExtractCompare):
+    @I.ir_module
+    class Before:
+        @R.function(pure=False)
+        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            @R.function(pure=False)
+            def inner_func(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+                z = R.add(x, y)
+                w = R.multiply(x, z)
+                v = R.add(y, w)
+                _ = R.print(format="oops")
+                a = R.multiply(v, v)
+                b = R.add(a, a)
+                c = R.multiply(a, b)
+                return c
+
+            z = R.add(x, y)
+            w = R.multiply(z, z)
+            v = R.divide(w, z)
+            q = inner_func(w, v)
+            a = R.multiply(q, q)
+            b = R.add(a, a)
+            c = R.multiply(b, a)
+            return c
+
+    @I.ir_module
+    class Expected:
+        @R.function(pure=False)
+        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            with R.dataflow():
+
+                @R.function(pure=False)
+                def inner_func(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+                    with R.dataflow():
+                        z = R.add(x, y)
+                        w = R.multiply(x, z)
+                        v = R.add(y, w)
+                        R.output(v)
+                    _ = R.print(format="oops")
+                    with R.dataflow():
+                        a = R.multiply(v, v)
+                        b = R.add(a, a)
+                        c = R.multiply(a, b)
+                        R.output(c)
+                    return c
+
+                z = R.add(x, y)
+                w = R.multiply(z, z)
+                v = R.divide(w, z)
+                R.output(inner_func, v, w)
+            q = inner_func(w, v)
+            with R.dataflow():
+                a = R.multiply(q, q)
+                b = R.add(a, a)
+                c = R.multiply(b, a)
+                R.output(c)
+            return c
+
```

Comment:
Can we add two test cases for making dataflow blocks when the individual streak doesn't satisfy the `min_size_` requirement, but the presence of a dataflow block next to the potentially-dataflow binding still avoids making a tiny dataflow block.

```python
class TestMergeWithPrecedingDataflowBlock(ExtractCompare):
    @I.ir_module
    class Before:
        @R.function
        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
            with R.dataflow():
                z = R.add(x, y)
                w = R.multiply(z, y)
                R.output(w)

            # The single binding of `v = R.add` would normally not be
            # enough to make a dataflow block, as `1 < min_size == 3`.
            v = R.add(w, x)
            return v

    @I.ir_module
    class Expected:
        @R.function
        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
            with R.dataflow():
                z = R.add(x, y)
                w = R.multiply(z, y)
                # However, it occurs just after an existing dataflow
                # block, and can be merged into it.
                v = R.add(w, x)
                R.output(v)
            return v


class TestMergeWithNextDataflowBlock(ExtractCompare):
    @I.ir_module
    class Before:
        @R.function
        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
            # The single binding of `z = R.add` would normally not be
            # enough to make a dataflow block, as `1 < min_size == 3`.
            z = R.add(x, y)

            # However, it occurs just before an existing dataflow
            # block, and can be merged into it.

            with R.dataflow():
                w = R.multiply(z, y)
                v = R.add(w, x)
                R.output(v)
            return v

    @I.ir_module
    class Expected:
        @R.function
        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
            with R.dataflow():
                z = R.add(x, y)
                w = R.multiply(z, y)
                v = R.add(w, x)
                R.output(v)
            return v
```

Reply:
Great idea. I'll have to handle that case though.

Reply:
Done, thank you